### PR TITLE
KAFKA-14367; Add `ListGroups` to the new `GroupCoordinator` interface

### PR DESCRIPTION
**Pull request by dajac**
This patch adds `listGroups` to the new `GroupCoordinator` interface and updates `KafkaApis` to use it.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (jolshan) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: same here with trace logging as the other PR.

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (jolshan) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: What was the reasoning for switching from the filter method to this approach?

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorAdapterTest.scala
Comment (jolshan) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorAdapterTest.scala:325]: Should we be including the empty group if the filter is stable?

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (dajac) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: I am actually not sure why I did that. Let me revert to the previous method.

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorAdapterTest.scala
Comment (dajac) [reply core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorAdapterTest.scala:325]: I agree that this is a tad weird. However, the filtering is not done by `GroupCoordinatorAdapter` so the test only verifies that inputs/outputs are correctly translated here. In practice, we would not have an error and a list of groups as well.

File: core/src/main/scala/kafka/server/KafkaApis.scala
Comment (jolshan) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: We will probably want to use the helper from the SyncGroup PR here too.

Comment (dajac) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Totally. Will do when that PR is merged.

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorAdapterTest.scala
Comment (jolshan) [reply core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorAdapterTest.scala:325]: Hmm I think I'm a bit confused by your last sentence. 
But to clarify my understanding -- this filtering is done within the handleListGroups method, but since this test is mocking the result we don't actually filter the groups? 

And the different filters at the top are just really checking the empty state filter is tested correctly?

**Review by jolshan — APPROVED** [review]
lgtm once we merge the other PR and add the helper.

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorAdapterTest.scala
Comment (dajac) [reply core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorAdapterTest.scala:325]: That’s right.

**Review by hachikuji — APPROVED** [review]
LGTM

**Pull request merged**
